File: packages/agenticcarekit/kernel/contracts/provider.py
"""Contract 1 — ``Capabilities`` and the ``Provider`` protocol.

Providers declare what they can do; the runtime negotiates. Capability
negotiation is the highest-leverage abstraction in the design: it turns
"audio is E2B/E4B only" from a doc footnote into a startup error with a
fix attached (invariant 2: never silently degrade).

Nothing here hides the provider — every concrete provider exposes its raw
client, and the kernel is callable directly (invariant 3: ejectable).
"""

from __future__ import annotations

from collections.abc import Iterator
from dataclasses import dataclass, field
from enum import StrEnum
from typing import Any, Literal, Protocol, runtime_checkable

__all__ = [
    "AudioPart",
    "Capabilities",
    "Chunk",
    "EgressClass",
    "GenerateRequest",
    "GenerateResponse",
    "ImageDetail",
    "ImagePart",
    "Message",
    "Modality",
    "Part",
    "Provider",
    "Role",
    "TextPart",
    "ToolCall",
    "Usage",
    "VISION_TOKEN_BUDGETS",
]


class Modality(StrEnum):
    """An input or output modality a model can consume or produce."""

    TEXT = "text"
    IMAGE = "image"
    AUDIO = "audio"


class EgressClass(StrEnum):
    """Where data goes when it reaches a provider. The privacy boundary
    (Contract 2) is defined in terms of these classes."""

    DEVICE = "device"                    # never leaves the machine
    TRUSTED_NETWORK = "trusted-network"  # self-hosted, user-controlled
    PUBLIC_CLOUD = "public-cloud"        # third-party API


#: Gemma 4 vision token budgets, exposed as named presets.
#: (quirk 5 in docs/brief.md §2 — raw budgets: 70, 140, 280, 560, 1120)
VISION_TOKEN_BUDGETS: dict[str, int] = {
    "minimal": 70,
    "caption": 140,
    "default": 280,
    "detail": 560,
    "ocr": 1120,
}

ImageDetail = Literal["minimal", "caption", "default", "detail", "ocr"]


@dataclass(frozen=True)
class Capabilities:
    """What a provider/model pair can do. Declared, never inferred.

    Example:
        >>> caps = Capabilities(
        ...     modalities_in=frozenset({Modality.TEXT, Modality.AUDIO}),
        ...     modalities_out=frozenset({Modality.TEXT}),
        ...     tool_calling=True, streaming=True,
        ...     context_tokens=131072, thinking=True,
        ...     egress=EgressClass.DEVICE,
        ... )
        >>> caps.missing(modalities_in=frozenset({Modality.IMAGE}))
        ['image input']
    """

    modalities_in: frozenset[Modality]
    modalities_out: frozenset[Modality]
    tool_calling: bool
    streaming: bool
    context_tokens: int
    thinking: bool
    egress: EgressClass

    def missing(
        self,
        *,
        modalities_in: frozenset[Modality] = frozenset(),
        modalities_out: frozenset[Modality] = frozenset(),
        tool_calling: bool = False,
        streaming: bool = False,
        context_tokens: int = 0,
        thinking: bool = False,
    ) -> list[str]:
        """Human-readable list of requirements this capability set lacks.

        Empty list means every requirement is met. The strings feed
        directly into ``CapabilityMismatch`` messages — they are part of
        the error contract, not decoration.
        """
        gaps: list[str] = []
        for m in sorted(modalities_in - self.modalities_in, key=lambda m: m.value):
            gaps.append(f"{m.value} input")
        for m in sorted(modalities_out - self.modalities_out, key=lambda m: m.value):
            gaps.append(f"{m.value} output")
        if tool_calling and not self.tool_calling:
            gaps.append("tool calling")
        if streaming and not self.streaming:
            gaps.append("streaming")
        if context_tokens > self.context_tokens:
            gaps.append(
                f"context window ({context_tokens} needed, {self.context_tokens} available)"
            )
        if thinking and not self.thinking:
            gaps.append("thinking")
        return gaps


# ── Messages ─────────────────────────────────────────────────────────────

Role = Literal["system", "user", "assistant", "tool"]


@dataclass(frozen=True)
class TextPart:
    text: str


@dataclass(frozen=True)
class ImagePart:
    """Image content. ``data`` is raw bytes or a filesystem path/base64 string.

    ``detail`` selects a Gemma 4 vision token budget preset
    (see ``VISION_TOKEN_BUDGETS``).
    """

    data: bytes | str
    detail: ImageDetail = "default"


@dataclass(frozen=True)
class AudioPart:
    """Audio content. ``data`` is raw bytes or a filesystem path/base64 string."""

    data: bytes | str
    format: str = "wav"


Part = TextPart | ImagePart | AudioPart


@dataclass(frozen=True)
class ToolCall:
    """A function call requested by the model (native function calling)."""

    id: str
    name: str
    arguments: dict[str, Any]


@dataclass(frozen=True)
class Message:
    """One conversation turn.

    ``thinking`` holds the model's thought block for assistant turns. The
    message builder strips prior-turn thought blocks from history
    automatically (quirk 3 — a silent correctness bug otherwise); keeping
    thinking OUT of ``parts`` is what makes that stripping structural
    rather than string surgery.
    """

    role: Role
    parts: tuple[Part, ...]
    thinking: str | None = None
    tool_calls: tuple[ToolCall, ...] = ()
    tool_call_id: str | None = None

    @staticmethod
    def text(role: Role, text: str) -> Message:
        """Convenience constructor for a plain text turn.

        Example:
            >>> Message.text("user", "hi").parts[0].text
            'hi'
        """
        return Message(role=role, parts=(TextPart(text),))

    def required_modalities(self) -> frozenset[Modality]:
        """The input modalities this message needs from a model."""
        mods = set()
        for p in self.parts:
            if isinstance(p, TextPart):
                mods.add(Modality.TEXT)
            elif isinstance(p, ImagePart):
                mods.add(Modality.IMAGE)
            elif isinstance(p, AudioPart):
                mods.add(Modality.AUDIO)
        return frozenset(mods)


# ── Requests and responses ───────────────────────────────────────────────


@dataclass(frozen=True)
class GenerateRequest:
    """A single generation request.

    Sampling fields default to ``None`` meaning "apply the model's known
    good defaults" (for Gemma 4: temperature=1.0, top_p=0.95, top_k=64 —
    quirk 1). Set a value only to override deliberately.

    ``think=True`` enables thinking (quirk 2: the provider injects the
    ``<|think|>`` token at the start of the system prompt).
    """

    messages: tuple[Message, ...]
    model: str | None = None
    tools: tuple[Any, ...] = ()  # tuple[ToolSpec, ...]; Any avoids an import cycle
    think: bool = False
    temperature: float | None = None
    top_p: float | None = None
    top_k: int | None = None
    max_tokens: int | None = None
    stop: tuple[str, ...] = ()

    def required_modalities(self) -> frozenset[Modality]:
        """Union of input modalities across all messages."""
        mods: frozenset[Modality] = frozenset()
        for m in self.messages:
            mods |= m.required_modalities()
        return mods


@dataclass(frozen=True)
class Usage:
    input_tokens: int = 0
    output_tokens: int = 0


@dataclass(frozen=True)
class GenerateResponse:
    """A completed generation. ``raw`` always carries the unmodified
    provider payload — the escape hatch that keeps this a toolkit, not a
    framework."""

    text: str
    thinking: str | None = None
    tool_calls: tuple[ToolCall, ...] = ()
    usage: Usage = field(default_factory=Usage)
    model: str = ""
    raw: dict[str, Any] = field(default_factory=dict)


@dataclass(frozen=True)
class Chunk:
    """One streaming increment. The final chunk has ``done=True`` and
    carries the assembled ``GenerateResponse``."""

    delta: str = ""
    thinking_delta: str = ""
    tool_call: ToolCall | None = None
    done: bool = False
    response: GenerateResponse | None = None


@runtime_checkable
class Provider(Protocol):
    """The provider contract. Concrete implementations live in
    ``agenticcarekit.kernel.providers`` (W-A); anything satisfying this
    protocol — including third-party plugins — plugs in identically.
    """

    name: str

    def capabilities(self) -> Capabilities: ...

    def generate(self, req: GenerateRequest) -> GenerateResponse: ...

    def stream(self, req: GenerateRequest) -> Iterator[Chunk]: ...

File: packages/ts/src/contracts/provider.ts
/**
 * Contract 1 — `Capabilities` and the `Provider` interface.
 *
 * Providers declare what they can do; the runtime negotiates. Capability
 * negotiation turns "audio is E2B/E4B only" from a doc footnote into a
 * startup error with a fix attached (invariant 2: never silently degrade).
 *
 * Nothing here hides the provider — concrete providers expose their raw
 * client by convention (`client`), and the kernel stays callable directly.
 */

/** An input or output modality a model can consume or produce. */
export const Modality = {
  TEXT: "text",
  IMAGE: "image",
  AUDIO: "audio",
} as const;
export type Modality = (typeof Modality)[keyof typeof Modality];

export const MODALITIES: readonly Modality[] = [Modality.TEXT, Modality.IMAGE, Modality.AUDIO];

export function isModality(value: string): value is Modality {
  return (MODALITIES as readonly string[]).includes(value);
}

/**
 * Where data goes when it reaches a provider. The privacy boundary
 * (Contract 2) is defined over these three classes and nothing else.
 */
export const EgressClass = {
  DEVICE: "device", // never leaves the machine
  TRUSTED_NETWORK: "trusted-network", // self-hosted, user-controlled
  PUBLIC_CLOUD: "public-cloud", // third-party API
} as const;
export type EgressClass = (typeof EgressClass)[keyof typeof EgressClass];

export const EGRESS_CLASSES: readonly EgressClass[] = [
  EgressClass.DEVICE,
  EgressClass.TRUSTED_NETWORK,
  EgressClass.PUBLIC_CLOUD,
];

export function isEgressClass(value: string): value is EgressClass {
  return (EGRESS_CLASSES as readonly string[]).includes(value);
}

/**
 * Gemma 4 vision token budgets, exposed as named presets
 * (docs/brief.md §2, quirk 5).
 */
export const VISION_TOKEN_BUDGETS: Readonly<Record<string, number>> = {
  minimal: 70,
  caption: 140,
  default: 280,
  detail: 560,
  ocr: 1120,
};

export type ImageDetail = "minimal" | "caption" | "default" | "detail" | "ocr";

export interface CapabilityRequirements {
  modalitiesIn?: Iterable<Modality>;
  modalitiesOut?: Iterable<Modality>;
  toolCalling?: boolean;
  streaming?: boolean;
  contextTokens?: number;
  thinking?: boolean;
}

export interface CapabilitiesInit {
  modalitiesIn: Iterable<Modality>;
  modalitiesOut: Iterable<Modality>;
  toolCalling: boolean;
  streaming: boolean;
  contextTokens: number;
  thinking: boolean;
  egress: EgressClass;
}

/**
 * What a provider/model pair can do. Declared, never inferred.
 *
 * @example
 * const caps = new Capabilities({
 *   modalitiesIn: ["text", "audio"], modalitiesOut: ["text"],
 *   toolCalling: true, streaming: true, contextTokens: 131072,
 *   thinking: true, egress: "device",
 * });
 * caps.missing({ modalitiesIn: ["image"] }); // ["image input"]
 */
export class Capabilities {
  readonly modalitiesIn: ReadonlySet<Modality>;
  readonly modalitiesOut: ReadonlySet<Modality>;
  readonly toolCalling: boolean;
  readonly streaming: boolean;
  readonly contextTokens: number;
  readonly thinking: boolean;
  readonly egress: EgressClass;

  constructor(init: CapabilitiesInit) {
    this.modalitiesIn = new Set(init.modalitiesIn);
    this.modalitiesOut = new Set(init.modalitiesOut);
    this.toolCalling = init.toolCalling;
    this.streaming = init.streaming;
    this.contextTokens = init.contextTokens;
    this.thinking = init.thinking;
    this.egress = init.egress;
  }

  /**
   * Human-readable list of requirements this capability set lacks.
   *
   * An empty list means every requirement is met. These strings feed
   * `CapabilityMismatch` messages verbatim — they are part of the error
   * contract, not decoration, and so is their order: input modalities
   * (sorted by name), then output modalities (sorted), then tool calling,
   * streaming, context window, thinking.
   */
  missing(requirements: CapabilityRequirements = {}): string[] {
    const gaps: string[] = [];
    const wantedIn = [...new Set(requirements.modalitiesIn ?? [])]
      .filter((m) => !this.modalitiesIn.has(m))
      .sort();
    for (const m of wantedIn) gaps.push(`${m} input`);

    const wantedOut = [...new Set(requirements.modalitiesOut ?? [])]
      .filter((m) => !this.modalitiesOut.has(m))
      .sort();
    for (const m of wantedOut) gaps.push(`${m} output`);

    if (requirements.toolCalling && !this.toolCalling) gaps.push("tool calling");
    if (requirements.streaming && !this.streaming) gaps.push("streaming");
    const contextTokens = requirements.contextTokens ?? 0;
    if (contextTokens > this.contextTokens) {
      gaps.push(`context window (${contextTokens} needed, ${this.contextTokens} available)`);
    }
    if (requirements.thinking && !this.thinking) gaps.push("thinking");
    return gaps;
  }

  /** The `capabilities.schema.json` document for this record. */
  toJSON(): Record<string, unknown> {
    return {
      modalities_in: [...this.modalitiesIn].sort(),
      modalities_out: [...this.modalitiesOut].sort(),
      tool_calling: this.toolCalling,
      streaming: this.streaming,
      context_tokens: this.contextTokens,
      thinking: this.thinking,
      egress: this.egress,
    };
  }
}

// ── Messages ─────────────────────────────────────────────────────────────

export type Role = "system" | "user" | "assistant" | "tool";

/** Media payload: raw bytes, or a string that is a path or already base64. */
export type MediaData = Uint8Array | string;

export interface TextPart {
  readonly type: "text";
  readonly text: string;
}

export interface ImagePart {
  readonly type: "image";
  readonly data: MediaData;
  readonly detail: ImageDetail;
}

export interface AudioPart {
  readonly type: "audio";
  readonly data: MediaData;
  readonly format: string;
}

export type Part = TextPart | ImagePart | AudioPart;

export function textPart(text: string): TextPart {
  return { type: "text", text };
}

export function imagePart(data: MediaData, detail: ImageDetail = "default"): ImagePart {
  return { type: "image", data, detail };
}

export function audioPart(data: MediaData, format = "wav"): AudioPart {
  return { type: "audio", data, format };
}

/** A function call requested by the model (native function calling). */
export interface ToolCall {
  readonly id: string;
  readonly name: string;
  readonly arguments: Record<string, unknown>;
}

export interface MessageInit {
  role: Role;
  parts?: readonly Part[];
  /**
   * The model's thought block for assistant turns. Kept OUT of `parts` so
   * that stripping prior-turn thinking (quirk 3) is structural rather than
   * string surgery — the builder simply never serializes this field.
   */
  thinking?: string | null;
  toolCalls?: readonly ToolCall[];
  toolCallId?: string | null;
}

/** One conversation turn. */
export class Message {
  readonly role: Role;
  readonly parts: readonly Part[];
  readonly thinking: string | null;
  readonly toolCalls: readonly ToolCall[];
  readonly toolCallId: string | null;

  constructor(init: MessageInit) {
    this.role = init.role;
    this.parts = init.parts ?? [];
    this.thinking = init.thinking ?? null;
    this.toolCalls = init.toolCalls ?? [];
    this.toolCallId = init.toolCallId ?? null;
  }

  /**
   * Convenience constructor for a plain text turn.
   *
   * @example
   * Message.text("user", "hi").parts[0]; // { type: "text", text: "hi" }
   */
  static text(role: Role, text: string): Message {
    return new Message({ role, parts: [textPart(text)] });
  }

  /** The input modalities this message needs from a model. */
  requiredModalities(): Set<Modality> {
    const mods = new Set<Modality>();
    for (const part of this.parts) {
      if (part.type === "text") mods.add(Modality.TEXT);
      else if (part.type === "image") mods.add(Modality.IMAGE);
      else mods.add(Modality.AUDIO);
    }
    return mods;
  }
}

// ── Requests and responses ───────────────────────────────────────────────

export interface ToolDeclaration {
  readonly name: string;
  readonly description: string;
  asFunctionSchema(): Record<string, unknown>;
}

export interface GenerateRequestInit {
  messages: readonly Message[];
  model?: string | null;
  tools?: readonly ToolDeclaration[];
  think?: boolean;
  temperature?: number | null;
  topP?: number | null;
  topK?: number | null;
  maxTokens?: number | null;
  stop?: readonly string[];
}

/**
 * A single generation request.
 *
 * Sampling fields default to `null` meaning "apply the model's known-good
 * defaults" (Gemma 4: 1.0 / 0.95 / 64 — quirk 1). Set a value only to
 * override deliberately; `0` is an override, not an absent value.
 */
export class GenerateRequest {
  readonly messages: readonly Message[];
  readonly model: string | null;
  readonly tools: readonly ToolDeclaration[];
  readonly think: boolean;
  readonly temperature: number | null;
  readonly topP: number | null;
  readonly topK: number | null;
  readonly maxTokens: number | null;
  readonly stop: readonly string[];

  constructor(init: GenerateRequestInit) {
    this.messages = init.messages;
    this.model = init.model ?? null;
    this.tools = init.tools ?? [];
    this.think = init.think ?? false;
    this.temperature = init.temperature ?? null;
    this.topP = init.topP ?? null;
    this.topK = init.topK ?? null;
    this.maxTokens = init.maxTokens ?? null;
    this.stop = init.stop ?? [];
  }

  /** Union of input modalities across all messages. */
  requiredModalities(): Set<Modality> {
    const mods = new Set<Modality>();
    for (const message of this.messages) {
      for (const m of message.requiredModalities()) mods.add(m);
    }
    return mods;
  }
}

export interface Usage {
  readonly inputTokens: number;
  readonly outputTokens: number;
}

export function usage(inputTokens = 0, outputTokens = 0): Usage {
  return { inputTokens, outputTokens };
}

/**
 * A completed generation. `raw` always carries the unmodified provider
 * payload — the escape hatch that keeps this a toolkit, not a framework.
 */
export interface GenerateResponse {
  readonly text: string;
  readonly thinking?: string | null;
  readonly toolCalls?: readonly ToolCall[];
  readonly usage?: Usage;
  readonly model?: string;
  readonly raw?: Record<string, unknown>;
}

/**
 * One streaming increment. The final chunk has `done: true` and carries the
 * assembled `GenerateResponse`.
 */
export interface Chunk {
  readonly delta?: string;
  readonly thinkingDelta?: string;
  readonly toolCall?: ToolCall | null;
  readonly done?: boolean;
  readonly response?: GenerateResponse | null;
}

/**
 * The provider contract. Anything satisfying it plugs in identically,
 * including third-party plugins.
 */
export interface Provider {
  readonly name: string;
  capabilities(): Capabilities;
  generate(req: GenerateRequest): GenerateResponse | Promise<GenerateResponse>;
  stream?(req: GenerateRequest): AsyncIterable<Chunk> | Iterable<Chunk>;
}
